<a href="https://colab.research.google.com/github/shreyasur/Data_Wrangling_Assignment/blob/main/Data_Wrangling2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import io
from google.colab import files

In [2]:
uploaded = files.upload()
patients = pd.read_csv(io.StringIO(uploaded['patients.csv'].decode('utf-8')))
patients

Saving patients.csv to patients.csv


patient_id assigned_sex given_name  ... weight height   bmi
0             1       female        Zoe  ...  121.7     66  19.6
1             2       female     Pamela  ...  118.8     66  19.2
2             3         male        Jae  ...  177.8     71  24.8
3             4         male       Liêm  ...  220.9     70  31.7
4             5         male        Tim  ...  192.3     27  26.1
..          ...          ...        ...  ...    ...    ...   ...
498         499         male    Mustafa  ...  181.1     72  24.6
499         500         male      Ruman  ...  239.6     70  34.4
500         501       female      Jinke  ...  171.2     67  26.8
501         502       female    Chidalu  ...  176.9     67  27.7
502         503         male        Pat  ...  138.2     71  19.3

[503 rows x 14 columns]

In [3]:
uploaded = files.upload()
treatment = pd.read_csv(io.StringIO(uploaded['treatments.csv'].decode('utf-8')))
treatment

Saving treatments.csv to treatments.csv


given_name       surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0     veronika      jindrová  41u - 48u  ...        7.63       7.20           NaN
1       elliot    richardson          -  ...        7.56       7.09          0.97
2     yukitaka      takenaka          -  ...        7.68       7.25           NaN
3         skye    gormanston  33u - 36u  ...        7.97       7.62          0.35
4       alissa        montez          -  ...        7.78       7.46          0.32
..         ...           ...        ...  ...         ...        ...           ...
275     albina      zetticci  45u - 51u  ...        7.93       7.73          0.20
276       john   teichelmann          -  ...        7.90       7.58           NaN
277     mathea       lillebø  23u - 36u  ...        9.04       8.67          0.37
278     vallie        prince  31u - 38u  ...        7.64       7.28          0.36
279     samúel  guðbrandsson  53u - 56u  ...        8.00       7.64          0.36

[280 rows x 7 columns]

In [4]:
uploaded = files.upload()
adverse = pd.read_csv(io.StringIO(uploaded['adverse_reactions.csv'].decode('utf-8')))

adverse

Saving adverse_reactions.csv to adverse_reactions.csv


given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [5]:
print(patients.shape)
print(treatment.shape)
print(adverse.shape)

(503, 14)
(280, 7)
(34, 3)


In [6]:
patients.head()

patient_id assigned_sex given_name  surname  ...  birthdate weight height   bmi
0           1       female        Zoe  Wellish  ...  7/10/1976  121.7     66  19.6
1           2       female     Pamela     Hill  ...   4/3/1967  118.8     66  19.2
2           3         male        Jae   Debord  ...  2/19/1980  177.8     71  24.8
3           4         male       Liêm     Phan  ...  7/26/1951  220.9     70  31.7
4           5         male        Tim  Neudorf  ...  2/18/1928  192.3     27  26.1

[5 rows x 14 columns]

In [7]:
patients.sample(5)

patient_id assigned_sex given_name  ... weight height   bmi
250         251       female        Mee  ...  206.1     64  35.4
339         340         male      Pavel  ...  229.0     67  35.9
216         217       female  Gabrielle  ...  170.1     67  26.6
286         287         male      Lewis  ...  155.3     68  23.6
91           92       female   Caroline  ...  218.7     66  35.3

[5 rows x 14 columns]

In [8]:
treatment.head()

given_name     surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0   veronika    jindrová  41u - 48u  ...        7.63       7.20           NaN
1     elliot  richardson          -  ...        7.56       7.09          0.97
2   yukitaka    takenaka          -  ...        7.68       7.25           NaN
3       skye  gormanston  33u - 36u  ...        7.97       7.62          0.35
4     alissa      montez          -  ...        7.78       7.46          0.32

[5 rows x 7 columns]

In [9]:
adverse.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

#####QUALITY

[ completeness, validity, accurate, consistency]

####1.PATIENTS
 - Zip code column is stored in decimal format [V] (done)
 - certain zipcodes are of 4 digits [V] (done)
 - for the state column sometimes full names are used and other times short forms are used [IC] (done)
 - Missing values for contact details [C] (done)
 - incorrect data types for assigned_sex and birthdate(done)
 - Errorneous data for weights(min weight = 48 pounds) and height (min height = 27 inches) [A] (done)
 - inconsistent way of storing phone numbers in contact column [IC]
 - duplicate and not useful data in dataset(needs to be removed) [IC] (done)

####2.TREATMENTS
- hba1c_change col has missing values(109 values) [C] (DONE)
- Novodra and Auralin columns has u attached to numbers [V] (DONE)
- only 280 rows are present, instead of 350 [C] (DONE)
- Name and surname column is in lowercase, different from the format in given in patients table(camel case) [IC] (DONE)
- incorrect data type for Auralin and Novodra [V] (DONE)
- in hba1c_change 9 instead of 4 [A] (DONE)
- use of dashes instead of NaN in aurolin and novodra columns[IC] (DONE)


#####3.ADVERSE REACTION

- name and surname in lower case [IC] (DONE)


In [10]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [11]:
patients.isnull().sum()

patient_id       0
assigned_sex     0
given_name       0
surname          0
address         12
city            12
state           12
zip_code        12
country         12
contact         12
birthdate        0
weight           0
height           0
bmi              0
dtype: int64

In [12]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name  ... weight height   bmi
209         210       female     Lalita  ...  143.4     62  26.2
219         220         male         Mỹ  ...  237.8     69  35.1
230         231       female  Elisabeth  ...  165.9     63  29.4
234         235       female    Martina  ...  199.5     65  33.2
242         243         male       John  ...  205.3     74  26.4
249         250         male   Benjamin  ...  146.5     69  21.6
257         258         male        Jin  ...  231.7     69  34.2
264         265       female   Wafiyyah  ...  158.6     63  28.1
269         270       female     Flavia  ...  175.2     61  33.1
278         279       female   Generosa  ...  124.3     69  18.4
286         287         male      Lewis  ...  155.3     68  23.6
296         297       female        Chỉ  ...  181.1     63  32.1

[12 rows x 14 columns]

In [13]:
patients.describe().T

count          mean           std  ...      50%       75%      max
patient_id  503.0    252.000000    145.347859  ...    252.0    377.50    503.0
zip_code    491.0  49084.118126  30265.807442  ...  48057.0  75679.00  99701.0
weight      503.0    173.434990     33.916741  ...    175.3    199.50    255.9
height      503.0     66.634195      4.411297  ...     67.0     70.00     79.0
bmi         503.0     27.483897      5.276438  ...     27.2     31.75     37.7

[5 rows x 8 columns]

In [14]:
patients.sample(5)

patient_id assigned_sex given_name  ... weight height   bmi
41           42         male     Fakhri  ...  143.0     66  23.1
394         395       female       Anja  ...  182.2     61  34.4
234         235       female    Martina  ...  199.5     65  33.2
79           80         male       Eniz  ...  183.9     72  24.9
137         138       female       Emma  ...  175.6     60  34.3

[5 rows x 14 columns]

In [15]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name  ... weight height   bmi
29           30         male       Jake  ...  155.8     67  24.4
219         220         male         Mỹ  ...  237.8     69  35.1
229         230         male       John  ...  180.0     72  24.4
230         231       female  Elisabeth  ...  165.9     63  29.4
234         235       female    Martina  ...  199.5     65  33.2
237         238         male       John  ...  180.0     72  24.4
242         243         male       John  ...  205.3     74  26.4
244         245         male       John  ...  180.0     72  24.4
249         250         male   Benjamin  ...  146.5     69  21.6
251         252         male       John  ...  180.0     72  24.4
257         258         male        Jin  ...  231.7     69  34.2
264         265       female   Wafiyyah  ...  158.6     63  28.1
269         270       female     Flavia  ...  175.2     61  33.1
277         278         male       John  ...  180.0     72  24.4
278         279       female   Generosa  ...  124.3     69  18.4
282         283       female      Sandy  ...  206.1     64  35.4
286         287         male      Lewis  ...  155.3     68  23.6
296         297       female        Chỉ  ...  181.1     63  32.1
502         503         male        Pat  ...  138.2     71  19.3

[19 rows x 14 columns]

In [16]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [17]:
treatment.isnull().sum()

given_name        0
surname           0
auralin           0
novodra           0
hba1c_start       0
hba1c_end         0
hba1c_change    109
dtype: int64

In [18]:
treatment.describe().T

count      mean       std   min   25%   50%   75%   max
hba1c_start   280.0  7.985929  0.568638  7.50  7.66  7.80  7.97  9.95
hba1c_end     280.0  7.589286  0.569672  7.01  7.27  7.42  7.57  9.58
hba1c_change  171.0  0.546023  0.279555  0.20  0.34  0.38  0.92  0.99

In [19]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [20]:
adverse.isnull().sum()

given_name          0
surname             0
adverse_reaction    0
dtype: int64

#### UNTIDY DATA

#### PATIENTS
- phone and email in one column i.e the contact column

#### TREATMENT
- 3 variables stored in auralin and novodra column(3 seperate columns should be there- lower dose,upper dose and treatment type)

#### ADVERSE
- no point of using this table , instead we can just add a column in the treatments table

#### A. COMPLETENESS ISSUES
- missing values for contact details of 12 patients (removed)
- only 280 rows are there instead of 350 (done)
- hba1c_change column has missing values(done)

In [21]:

# creating copies

patients_copy=patients.copy()
treatment_copy=treatment.copy()
adverse_copy=adverse.copy()

DEFINE

In [22]:
# SUBTRACT hba1c_start from hba1c_end to get hba1c_change


CODE

In [23]:
treatment_copy['hba1c_change']=treatment_copy['hba1c_start']-treatment_copy['hba1c_end']


TEST

In [24]:
treatment_copy.isnull().sum()

given_name      0
surname         0
auralin         0
novodra         0
hba1c_start     0
hba1c_end       0
hba1c_change    0
dtype: int64

#### B.TIDINESS ISSUES

- phone + email in col (patients) (seperated them)
- Novodra and Auolin col treatment ( done)
- Merge the adverse table to treatment table (done)


DEFINE

In [25]:
# Novodra and Aurolin using the melt function

treatment_copy=pd.melt(treatment_copy , id_vars=['given_name','surname','hba1c_start','hba1c_end','hba1c_change'], var_name='treatment',value_name='dose')

In [26]:
treatment_copy=treatment_copy[treatment_copy['dose']!='-']

In [27]:
treatment_copy['start'],treatment_copy['end']=treatment_copy['dose'].str.split('-').str

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  """Entry point for launching an IPython kernel.


In [28]:
treatment_copy.drop('dose',axis=1,inplace=True)

TEST

In [29]:
treatment_copy.head()

given_name     surname  hba1c_start  ...  treatment  start   end
0   veronika    jindrová         7.63  ...    auralin   41u    48u
3       skye  gormanston         7.97  ...    auralin   33u    36u
6     sophia      haugen         7.65  ...    auralin   37u    42u
7      eddie      archer         7.89  ...    auralin   31u    38u
9       asia     woźniak         7.76  ...    auralin   30u    36u

[5 rows x 8 columns]

ADDING THE MISSING ROWS IN TREATMENTS COLUMN



In [30]:
uploaded = files.upload()
treatment_cut = pd.read_csv(io.StringIO(uploaded['treatments_cut.csv'].decode('utf-8')))
treatment_cut

Saving treatments_cut.csv to treatments_cut.csv


given_name     surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0       jožka   resanovič  22u - 30u  ...        7.56       7.22          0.34
1   inunnguaq    heilmann  57u - 67u  ...        7.85       7.45           NaN
2       alwin    svensson  36u - 39u  ...        7.78       7.34           NaN
3         thể       lương          -  ...        7.64       7.22          0.92
4      amanda     ribeiro  36u - 44u  ...        7.85       7.47          0.38
..        ...         ...        ...  ...         ...        ...           ...
65     rovzan     kishiev  32u - 37u  ...        7.75       7.41          0.34
66      jakob    jakobsen          -  ...        7.96       7.51          0.95
67      bernd   schneider  48u - 56u  ...        7.74       7.44          0.30
68      berta  napolitani          -  ...        7.68       7.21           NaN
69     armina       sauvé  36u - 46u  ...        7.86       7.40           NaN

[70 rows x 7 columns]

In [31]:
treatment_copy.shape

(280, 8)

In [32]:
# maintaining a copy of treatment_cut
treatment_cut_copy=treatment_cut.copy()

In [33]:
treatment_cut_copy.head()

given_name    surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0      jožka  resanovič  22u - 30u  ...        7.56       7.22          0.34
1  inunnguaq   heilmann  57u - 67u  ...        7.85       7.45           NaN
2      alwin   svensson  36u - 39u  ...        7.78       7.34           NaN
3        thể      lương          -  ...        7.64       7.22          0.92
4     amanda    ribeiro  36u - 44u  ...        7.85       7.47          0.38

[5 rows x 7 columns]

PREPROCESSING THE TREATMENT_CUT SIMILAR TO TREATMENT

DEFINE

- THERE ARE MISSING VALUES IN hba1c_change
- CONVERTED THE AURALIN AND NOVODRA COLUMN INTO START END AND TREATMENT

CODE

In [34]:
treatment_cut_copy['hba1c_change']=treatment_cut_copy['hba1c_start']-treatment_cut_copy['hba1c_end']


In [35]:
treatment_cut_copy.isnull().sum()

given_name      0
surname         0
auralin         0
novodra         0
hba1c_start     0
hba1c_end       0
hba1c_change    0
dtype: int64

In [36]:
treatment_cut_copy=pd.melt(treatment_cut_copy , id_vars=['given_name','surname','hba1c_start','hba1c_end','hba1c_change'], var_name='treatment',value_name='dose')

In [37]:
treatment_cut_copy.head()

given_name    surname  hba1c_start  ...  hba1c_change  treatment       dose
0      jožka  resanovič         7.56  ...          0.34    auralin  22u - 30u
1  inunnguaq   heilmann         7.85  ...          0.40    auralin  57u - 67u
2      alwin   svensson         7.78  ...          0.44    auralin  36u - 39u
3        thể      lương         7.64  ...          0.42    auralin          -
4     amanda    ribeiro         7.85  ...          0.38    auralin  36u - 44u

[5 rows x 7 columns]

In [38]:
treatment_cut_copy=treatment_cut_copy[treatment_cut_copy['dose']!='-']

In [39]:
treatment_cut_copy['start'],treatment_cut_copy['end']=treatment_cut_copy['dose'].str.split('-').str

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  """Entry point for launching an IPython kernel.


In [40]:
treatment_cut_copy.drop('dose',axis=1,inplace=True)

TEST

In [41]:
treatment_cut_copy.sample(5)

given_name       surname  hba1c_start  ...  treatment  start   end
11    gabriela  michalčíková         8.37  ...    auralin   26u    36u
46        anja       mueller         7.58  ...    auralin   43u    56u
124      borna     šlezinger         7.55  ...    novodra   42u    41u
35       alvin       jackson         7.62  ...    auralin   38u    43u
103      meaza        brhane         7.70  ...    novodra   37u    41u

[5 rows x 8 columns]

MERGING TREATMENT AND TREATMENT CUT

- THERE ARE MISSING ROWS PRESENT IN TREATMENT
- 350 ROWS ARE SUPPOSED TO BE PRESENT WHEREAS ONLY 280 ARE PRESENT

In [42]:
treatment_copy=pd.concat([treatment_copy,treatment_cut_copy])

TEST

In [43]:
treatment_copy.shape

(350, 8)

In [44]:
treatment_copy.tail()

given_name     surname  hba1c_start  ...  treatment  start   end
128  christopher    woodward         7.51  ...    novodra   55u    51u
130        maret    sultygov         7.67  ...    novodra   26u    23u
134        lixue       hsueh         9.21  ...    novodra   22u    23u
136        jakob    jakobsen         7.96  ...    novodra   28u    26u
138        berta  napolitani         7.68  ...    novodra   42u    44u

[5 rows x 8 columns]

CAPITILISING THE FIRST LETTER OF THE GIVEN_NAME ANS SURNAME 
- SIMILAR TO THE PATIENTS TABLE

In [45]:
def capital_given_name(value):
   
   return value.title()

In [46]:
capital_given_name('maret')

'Maret'

In [47]:
treatment_copy['given_name']=treatment_copy['given_name'].apply(capital_given_name)

In [48]:
treatment_copy['surname']=treatment_copy['surname'].apply(capital_given_name)

In [49]:
treatment_copy

given_name     surname  hba1c_start  ...  treatment  start   end
0       Veronika    Jindrová         7.63  ...    auralin   41u    48u
3           Skye  Gormanston         7.97  ...    auralin   33u    36u
6         Sophia      Haugen         7.65  ...    auralin   37u    42u
7          Eddie      Archer         7.89  ...    auralin   31u    38u
9           Asia     Woźniak         7.76  ...    auralin   30u    36u
..           ...         ...          ...  ...        ...    ...   ...
128  Christopher    Woodward         7.51  ...    novodra   55u    51u
130        Maret    Sultygov         7.67  ...    novodra   26u    23u
134        Lixue       Hsueh         9.21  ...    novodra   22u    23u
136        Jakob    Jakobsen         7.96  ...    novodra   28u    26u
138        Berta  Napolitani         7.68  ...    novodra   42u    44u

[350 rows x 8 columns]

REMOVING u FROM START AND END

In [50]:
treatment_copy['start']=treatment_copy['start'].str.split('u').str[0]

In [51]:
treatment_copy.head()

given_name     surname  hba1c_start  ...  treatment  start   end
0   Veronika    Jindrová         7.63  ...    auralin     41   48u
3       Skye  Gormanston         7.97  ...    auralin     33   36u
6     Sophia      Haugen         7.65  ...    auralin     37   42u
7      Eddie      Archer         7.89  ...    auralin     31   38u
9       Asia     Woźniak         7.76  ...    auralin     30   36u

[5 rows x 8 columns]

In [52]:
treatment_copy['end']=treatment_copy['end'].str.split('u').str[0]

In [53]:
treatment_copy.head()

given_name     surname  hba1c_start  ...  treatment  start  end
0   Veronika    Jindrová         7.63  ...    auralin     41   48
3       Skye  Gormanston         7.97  ...    auralin     33   36
6     Sophia      Haugen         7.65  ...    auralin     37   42
7      Eddie      Archer         7.89  ...    auralin     31   38
9       Asia     Woźniak         7.76  ...    auralin     30   36

[5 rows x 8 columns]

In [54]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 138
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   hba1c_start   350 non-null    float64
 3   hba1c_end     350 non-null    float64
 4   hba1c_change  350 non-null    float64
 5   treatment     350 non-null    object 
 6   start         350 non-null    object 
 7   end           350 non-null    object 
dtypes: float64(3), object(5)
memory usage: 24.6+ KB


CHANGING THE DTYPE OF START AND END
- FROM OBJECT WE ARE CHANGING IT TO INT32

In [55]:
treatment_copy['start']=treatment_copy['start'].astype('int32')
treatment_copy['end']=treatment_copy['end'].astype('int32')

In [56]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 138
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   hba1c_start   350 non-null    float64
 3   hba1c_end     350 non-null    float64
 4   hba1c_change  350 non-null    float64
 5   treatment     350 non-null    object 
 6   start         350 non-null    int32  
 7   end           350 non-null    int32  
dtypes: float64(3), int32(2), object(3)
memory usage: 21.9+ KB


In [57]:
adverse_copy.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [58]:
adverse_copy['given_name']=adverse_copy['given_name'].apply(capital_given_name)

In [59]:
adverse_copy['surname']=adverse_copy['surname'].apply(capital_given_name)

In [60]:
adverse_copy


given_name     surname           adverse_reaction
0         Berta  Napolitani  injection site discomfort
1          Lena        Baer               hypoglycemia
2        Joseph         Day               hypoglycemia
3        Flavia  Fiorentino                      cough
4       Manouck     Wubbels          throat irritation
5       Jasmine       Sykes               hypoglycemia
6        Louise     Johnson               hypoglycemia
7       Albinca     Komavec               hypoglycemia
8           Noe      Aranda               hypoglycemia
9         Sofia   Hermansen  injection site discomfort
10        Tegan     Johnson                   headache
11         Abel     Yonatan                      cough
12    Abdul-Nur         Isa               hypoglycemia
13         Leon      Scholz  injection site discomfort
14     Gabriele     Saenger               hypoglycemia
15       Jia Li        Teng                     nausea
16        Jakob    Jakobsen               hypoglycemia
17  Christopher    Woodward                     nausea
18          Ole    Petersen               hypoglycemia
19       Finley    Chandler                   headache
20     Anenechi       Chidi               hypoglycemia
21     Miłosław  Wiśniewski  injection site discomfort
22        Lixue       Hsueh  injection site discomfort
23        Merci      Leroux               hypoglycemia
24         Kang         Mai  injection site discomfort
25       Elliot  Richardson               hypoglycemia
26      Clinton      Miller          throat irritation
27       Idalia       Moore               hypoglycemia
28       Xiuxiu       Chang               hypoglycemia
29         Alex    Crawford               hypoglycemia
30       Monika      Lončar               hypoglycemia
31       Steven         Roy                   headache
32      Cecilie      Nilsen               hypoglycemia
33    Krisztina      Magyar               hypoglycemia

REMOVING DUPLICATED DATA

- treatment table has 1 duplicated value where the name is Joseph Day
- patient table has duplicated values where the name is John Doe
- adverse table has no duplicated values

In [61]:
treatment_copy.duplicated(['given_name','surname']).sum()

1

In [62]:
treatment_copy[treatment_copy.duplicated(['given_name','surname'])]

given_name surname  hba1c_start  ...  treatment  start end
136     Joseph     Day          7.7  ...    auralin     29  36

[1 rows x 8 columns]

In [63]:
treatment_copy[treatment_copy['given_name']=='Joseph']

given_name surname  hba1c_start  ...  treatment  start end
10      Joseph     Day         7.70  ...    auralin     29  36
136     Joseph     Day         7.70  ...    auralin     29  36
57      Joseph  Tucker         7.67  ...    auralin     48  56

[3 rows x 8 columns]

In [64]:
treatment_copy1=treatment_copy.copy()

In [65]:
treatment_copy1

given_name     surname  hba1c_start  ...  treatment  start end
0       Veronika    Jindrová         7.63  ...    auralin     41  48
3           Skye  Gormanston         7.97  ...    auralin     33  36
6         Sophia      Haugen         7.65  ...    auralin     37  42
7          Eddie      Archer         7.89  ...    auralin     31  38
9           Asia     Woźniak         7.76  ...    auralin     30  36
..           ...         ...          ...  ...        ...    ...  ..
128  Christopher    Woodward         7.51  ...    novodra     55  51
130        Maret    Sultygov         7.67  ...    novodra     26  23
134        Lixue       Hsueh         9.21  ...    novodra     22  23
136        Jakob    Jakobsen         7.96  ...    novodra     28  26
138        Berta  Napolitani         7.68  ...    novodra     42  44

[350 rows x 8 columns]

In [66]:
treatment_copy.drop_duplicates(subset=['given_name','surname'],inplace=True)

In [67]:
treatment_copy[treatment_copy['given_name']=='Joseph']

given_name surname  hba1c_start  ...  treatment  start end
10     Joseph     Day         7.70  ...    auralin     29  36
57     Joseph  Tucker         7.67  ...    auralin     48  56

[2 rows x 8 columns]

In [68]:
treatment_copy.shape

(349, 8)

In [69]:
patients_copy.duplicated(['given_name','surname']).sum()

5

In [70]:
patients_copy[patients_copy.duplicated(['given_name','surname'])]

patient_id assigned_sex given_name surname  ... birthdate weight height   bmi
229         230         male       John     Doe  ...  1/1/1975  180.0     72  24.4
237         238         male       John     Doe  ...  1/1/1975  180.0     72  24.4
244         245         male       John     Doe  ...  1/1/1975  180.0     72  24.4
251         252         male       John     Doe  ...  1/1/1975  180.0     72  24.4
277         278         male       John     Doe  ...  1/1/1975  180.0     72  24.4

[5 rows x 14 columns]

In [71]:
patients_copy.drop_duplicates(subset=['given_name','surname'],inplace=True)

In [72]:
patients_copy[patients_copy['given_name']=='John']

patient_id assigned_sex given_name  ... weight height   bmi
140         141         male       John  ...  214.1     72  29.0
153         154         male       John  ...  177.8     68  27.0
215         216         male       John  ...  180.0     72  24.4
242         243         male       John  ...  205.3     74  26.4

[4 rows x 14 columns]

In [73]:
adverse_copy.duplicated(['given_name','surname']).sum()

0

DROPPING MISSING VALUES OF 12 ENTRIES IN PATIENTS DATAFRAME

In [74]:
patients_copy.isnull().sum()

patient_id       0
assigned_sex     0
given_name       0
surname          0
address         12
city            12
state           12
zip_code        12
country         12
contact         12
birthdate        0
weight           0
height           0
bmi              0
dtype: int64

In [75]:
patients_copy.dropna(subset=['address','city', 'state','zip_code','country','contact'],inplace=True)

In [76]:
patients_copy.isnull().sum()

patient_id      0
assigned_sex    0
given_name      0
surname         0
address         0
city            0
state           0
zip_code        0
country         0
contact         0
birthdate       0
weight          0
height          0
bmi             0
dtype: int64

In [77]:
patients_copy.shape

(486, 14)

In [78]:
treatment_copy.isnull().sum()

given_name      0
surname         0
hba1c_start     0
hba1c_end       0
hba1c_change    0
treatment       0
start           0
end             0
dtype: int64

In [79]:
adverse_copy.isnull().sum()

given_name          0
surname             0
adverse_reaction    0
dtype: int64

In [80]:
patients_copy.sample(5)

patient_id assigned_sex given_name  ... weight height   bmi
85           86         male    Burrell  ...  203.5     71  28.4
280         281         male      David  ...  128.5     67  20.1
468         469         male       Tosh  ...  194.7     68  29.6
104         105       female     Yasmin  ...  200.6     62  36.7
439         440         male     Hiromu  ...  242.0     70  34.7

[5 rows x 14 columns]

REPLACING THE DISCREPANCY IN THE STATE COLUMN
- SOME STATE ARE GIVEN IN FULL FORM OTHERS IN ABBREVIATION
- STATES WHOSE FULL FORM IS GIVEN INSTEAD OF ABBREVIATION
  - CALIFORINA-CA
  - NEW YORK - NY
  - ILLIONIS- IL
  - FLORIDA-FL
  - NEBRASKA - NE

In [81]:
patients_copy['state'].value_counts()

California    36
TX            32
New York      25
CA            24
MA            22
PA            18
NY            17
GA            15
OH            14
Illinois      14
OK            13
LA            13
Florida       13
MI            13
NJ            12
VA            11
IL            10
WI            10
MS            10
AL             9
TN             9
FL             9
IN             9
MN             9
KY             8
NC             8
WA             8
MO             7
NV             6
ID             6
KS             6
IA             5
SC             5
CT             5
CO             4
RI             4
Nebraska       4
AZ             4
AR             4
ME             4
ND             4
MD             3
SD             3
OR             3
DE             3
WV             3
VT             2
NE             2
DC             2
MT             2
AK             1
NH             1
WY             1
NM             1
Name: state, dtype: int64

In [82]:
patients_copy['state']=patients_copy['state'].replace('California','CA')
patients_copy['state']=patients_copy['state'].replace('New York','NY')
patients_copy['state']=patients_copy['state'].replace('Illinois','IL')
patients_copy['state']=patients_copy['state'].replace('Florida','FL')
patients_copy['state']=patients_copy['state'].replace('Nebraska','NE')


In [83]:
patients_copy['state'].value_counts()

CA    60
NY    42
TX    32
IL    24
FL    22
MA    22
PA    18
GA    15
OH    14
OK    13
MI    13
LA    13
NJ    12
VA    11
WI    10
MS    10
IN     9
MN     9
TN     9
AL     9
KY     8
NC     8
WA     8
MO     7
NE     6
KS     6
NV     6
ID     6
IA     5
CT     5
SC     5
ME     4
RI     4
ND     4
CO     4
AR     4
AZ     4
DE     3
OR     3
MD     3
WV     3
SD     3
DC     2
VT     2
MT     2
WY     1
AK     1
NM     1
NH     1
Name: state, dtype: int64

In [84]:
patients_copy.sample(5)

patient_id assigned_sex given_name  ... weight height   bmi
156         157         male      Thống  ...  172.9     68  26.3
322         323         male       Mười  ...  218.5     66  35.3
406         407         male     Dániel  ...  175.6     73  23.2
138         139         male       Jose  ...  171.2     74  22.0
193         194         male      Borna  ...  203.9     69  30.1

[5 rows x 14 columns]

CHANGING THE DTYPE OF ZIP_CODE FROM FLOAT TO INT

In [85]:
patients_copy['zip_code']=patients_copy['zip_code'].astype('int64')

SOLVING THE DISCREPANCY IN ZIP_CODE COLUMN
- SOME ZIP_CODES ARE GIVEN IN 4 DIGITS OTHER IN FIVE DIGITS
  - padding with 0 on the left where zip code is given in 4 digits
  - since zip codes starting with 0 are dropped
  - example - zip code of West Haven (CT) is given as 6516 but the zip code is 06516

In [86]:
patients_copy['zip_code']=patients_copy['zip_code'].astype('str').str.pad(5,'left','0')

In [87]:
patients_copy.sample(5)

patient_id assigned_sex given_name  ... weight height   bmi
144         145         male       Mile  ...  244.9     71  34.2
112         113       female  Arnþrúður  ...  172.0     63  30.5
412         413         male     Finlay  ...  161.3     70  23.1
162         163       female     Sarina  ...  201.7     63  35.7
392         393       female    Kristin  ...  206.1     67  32.3

[5 rows x 14 columns]

CHANGING ERRORNEOUS DATA PRESENT IN WEIGHT AND HEIGHT COLUMN

- THE MINIMUM HEIGHT IS GIVEN TO BE 27 INCHES, WITH  A WEIGHT OF 192.3 lbs AND BMI OF 26.1
- THE MINIMUM WEIGHT IS GIVEN TO BE 48.8 lbs, WITH A HEIGHT OF 63 INCHES AND BMI 0F 19.1 

In [88]:
patients_copy[patients_copy['weight'] == 48.8]

patient_id assigned_sex given_name  ... weight height   bmi
210         211       female    Camilla  ...   48.8     63  19.1

[1 rows x 14 columns]

In [89]:
patients_copy.describe().T

count        mean         std  ...     50%      75%    max
patient_id  486.0  251.991770  147.806895  ...  253.50  381.750  503.0
weight      486.0  173.278189   34.080077  ...  174.45  199.725  255.9
height      486.0   66.588477    4.414058  ...   67.00   70.000   79.0
bmi         486.0   27.491770    5.295547  ...   27.20   31.700   37.7

[4 rows x 8 columns]

In [90]:
patients_copy[patients_copy['height'] == 27.0]

patient_id assigned_sex given_name  surname  ...  birthdate weight height   bmi
4           5         male        Tim  Neudorf  ...  2/18/1928  192.3     27  26.1

[1 rows x 14 columns]

FORMULA FOR CALCULATING BMI-

- 703 x ((WEIGHT)/(HEIGHT)^2)

In [91]:
# CALCULATING HEIGHT FOR WEIGHT=192.3 AND BMI=26.1
#  THE HEIGHT COMES AROUND 71.96 ~ 72 INCHES


In [92]:
# UPDATING IT IN PATIENT_ID 5

patients_copy['height']=patients_copy['height'].replace(27,72)

TEST

In [93]:
patients_copy[patients_copy['patient_id'] == 5]

patient_id assigned_sex given_name  surname  ...  birthdate weight height   bmi
4           5         male        Tim  Neudorf  ...  2/18/1928  192.3     72  26.1

[1 rows x 14 columns]

In [94]:
patients_copy.describe().T

count        mean         std  ...     50%      75%    max
patient_id  486.0  251.991770  147.806895  ...  253.50  381.750  503.0
weight      486.0  173.278189   34.080077  ...  174.45  199.725  255.9
height      486.0   66.681070    4.037853  ...   67.00   70.000   79.0
bmi         486.0   27.491770    5.295547  ...   27.20   31.700   37.7

[4 rows x 8 columns]

In [95]:
# CALCULATING WEIGHT FOR HEIGHT=63 AND BMI=19.1
#  THE WEIGHT COMES AROUND 107.83


In [96]:
# UPDATING IT IN PATIENT ID 211

patients_copy['weight']=patients_copy['weight'].replace(48.8,107.8)

TEST

In [97]:
patients_copy[patients_copy['patient_id'] == 211]

patient_id assigned_sex given_name  ... weight height   bmi
210         211       female    Camilla  ...  107.8     63  19.1

[1 rows x 14 columns]

In [98]:
patients_copy.describe().T

count        mean         std  ...     50%      75%    max
patient_id  486.0  251.991770  147.806895  ...  253.50  381.750  503.0
weight      486.0  173.399588   33.739128  ...  174.45  199.725  255.9
height      486.0   66.681070    4.037853  ...   67.00   70.000   79.0
bmi         486.0   27.491770    5.295547  ...   27.20   31.700   37.7

[4 rows x 8 columns]

CHANGING THE DTYPE OF ASSIGNED_SEX AND BIRTHDATE

- ASSIGNED_SEX IS GIVEN AS OBJECT CHANGING IT TO CATEGORY
- BIRTHDATE IS GIVEN AS OBJECT CHANGING IT TO DATE

In [99]:
patients_copy['birthdate']=patients['birthdate']

In [100]:
patients_copy['birthdate']= pd.to_datetime(patients_copy['birthdate'])

In [101]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 486 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    486 non-null    int64         
 1   assigned_sex  486 non-null    object        
 2   given_name    486 non-null    object        
 3   surname       486 non-null    object        
 4   address       486 non-null    object        
 5   city          486 non-null    object        
 6   state         486 non-null    object        
 7   zip_code      486 non-null    object        
 8   country       486 non-null    object        
 9   contact       486 non-null    object        
 10  birthdate     486 non-null    datetime64[ns]
 11  weight        486 non-null    float64       
 12  height        486 non-null    int64         
 13  bmi           486 non-null    float64       
dtypes: datetime64[ns](1), float64(2), int64(2), object(9)
memory usage: 57.0+ KB


In [102]:
patients_copy['assigned_sex']=patients_copy['assigned_sex'].astype('category')

In [103]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 486 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    486 non-null    int64         
 1   assigned_sex  486 non-null    category      
 2   given_name    486 non-null    object        
 3   surname       486 non-null    object        
 4   address       486 non-null    object        
 5   city          486 non-null    object        
 6   state         486 non-null    object        
 7   zip_code      486 non-null    object        
 8   country       486 non-null    object        
 9   contact       486 non-null    object        
 10  birthdate     486 non-null    datetime64[ns]
 11  weight        486 non-null    float64       
 12  height        486 non-null    int64         
 13  bmi           486 non-null    float64       
dtypes: category(1), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.7+ KB


In [104]:
patients_copy.sample(5)

patient_id assigned_sex    given_name  ... weight height   bmi
333         334         male     Jad Allah  ...  174.2     69  25.7
479         480         male          Tapa  ...  220.0     65  36.6
217         218         male    Saidmohmad  ...  216.5     67  33.9
68           69       female        Haruna  ...  149.8     63  26.5
53           54         male  Kwemtochukwu  ...  150.5     72  20.4

[5 rows x 14 columns]

CHANGING THE FORMAT OF DATE


In [105]:
patients_copy['birthdate'] = patients_copy['birthdate'].dt.strftime('%d/%m/%Y')

In [106]:
patients_copy.head()

patient_id assigned_sex given_name  ... weight height   bmi
0           1       female        Zoe  ...  121.7     66  19.6
1           2       female     Pamela  ...  118.8     66  19.2
2           3         male        Jae  ...  177.8     71  24.8
3           4         male       Liêm  ...  220.9     70  31.7
4           5         male        Tim  ...  192.3     72  26.1

[5 rows x 14 columns]

In [124]:
#patients_copy.drop('DOB1',axis=1,inplace=True)

KeyError: ignored

ADDING THE ADVERSE_REACTION COLUMN FROM ADVERSE TO TREATMENT

In [133]:
treatment_copy= pd.merge(treatment_copy, adverse_copy, how="left", on=['given_name','surname'])

In [131]:
#treatment_copy.drop(['adverse_reaction_x','adverse_reaction_y'],axis=1,inplace=True)

In [135]:
treatment_copy.sample(5)

given_name     surname  hba1c_start  ...  start  end adverse_reaction
230     Martha  Afanasyeva         7.81  ...     34   41              NaN
262       Chen         Yao         7.90  ...     56   57              NaN
92       Merci      Leroux         8.98  ...     27   33     hypoglycemia
267     Csilla    Herczegh         7.71  ...     43   46              NaN
19    Kristján     Ingason         7.92  ...     44   55              NaN

[5 rows x 9 columns]

In [ ]:
#treatment_copy.drop('adverse_reaction')

In [136]:
adverse_copy[adverse_copy['given_name']=='Manouck']

given_name  surname   adverse_reaction
4    Manouck  Wubbels  throat irritation

In [137]:
treatment_copy.isnull().sum()

given_name            0
surname               0
hba1c_start           0
hba1c_end             0
hba1c_change          0
treatment             0
start                 0
end                   0
adverse_reaction    315
dtype: int64

SINCE IT IS NOT IMPORTANT WHAT TYPE OF ADVERSE REACTION WAS CAUSED,RATHER IT IS IMPORTANT IF ANY ADVERSE REACTION WAS CAUSED, THEREFORE CHANGING ADVERSE_REACTION INTO A CATEGORICAL COLUMN WITH N/Y AS VALUES

In [138]:
treatment_copy['adverse_reaction']=treatment_copy['adverse_reaction'].fillna('N')

In [139]:
def adverse_reaction(value):
  if value=='N':
    return 'N'
  else:
    return 'Y'

In [140]:
adverse_reaction('throat irritation')

'Y'

In [141]:
treatment_copy['adverse_reaction']=treatment_copy['adverse_reaction'].apply(adverse_reaction)

In [142]:
treatment_copy.sample(5)

given_name   surname  hba1c_start  ...  start  end adverse_reaction
103    Vanessa  Ferguson         7.97  ...     39   50                N
281      Alwin  Svensson         7.78  ...     36   39                N
99     Nicolas  Ferreira         7.99  ...     43   51                N
239      Najla     Touma         7.66  ...     39   44                N
251     Pamela      Hill         9.53  ...     27   29                N

[5 rows x 9 columns]

In [143]:
treatment_copy.isnull().sum()

given_name          0
surname             0
hba1c_start         0
hba1c_end           0
hba1c_change        0
treatment           0
start               0
end                 0
adverse_reaction    0
dtype: int64

In [ ]:
treatment_copy.info()

In [144]:
treatment_copy['adverse_reaction']=treatment_copy['adverse_reaction'].astype('category')

In [145]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   given_name        349 non-null    object  
 1   surname           349 non-null    object  
 2   hba1c_start       349 non-null    float64 
 3   hba1c_end         349 non-null    float64 
 4   hba1c_change      349 non-null    float64 
 5   treatment         349 non-null    object  
 6   start             349 non-null    int32   
 7   end               349 non-null    int32   
 8   adverse_reaction  349 non-null    category
dtypes: category(1), float64(3), int32(2), object(3)
memory usage: 22.2+ KB


SPLITTING THE CONTACT COLUMN OF PATIENTS -
- INTO EMAIL AND PHONE


In [ ]:
patients_copy.head()

In [ ]:
#patients_copy['contact'].str.split()

In [ ]:
 #patients_copy['contact'].str.extract(r'([\d:,' ']+)')

In [148]:
patients_copy['Email_address']=patients_copy['contact'].str.findall('[a-zA-Z_.]+@[a-zA-Z0-9-]+\.[a-zA-Z.]+').str[0]

In [149]:
patients_copy.sample(5)

patient_id assigned_sex  ...    phone_number                 Email_address
393         394         male  ...  1 703 716 8652          StevenBRoy@rhyta.com
6             7       female  ...    775-533-5933         MaryBAdams@einrot.com
30           31       female  ...    303-910-2058  JaninaBazarova@superrito.com
438         439       female  ...    330-365-4059  HannahMcIntosh@superrito.com
23           24         male  ...  1 813 355 9476          LovreGalic@gustr.com

[5 rows x 16 columns]

In [150]:
 patients_copy['phone_number']=patients_copy['contact'].str.findall(r'[\+\(]?[1-9][0-9 .\-\(\)]{8,}[0-9]').str[0]

In [151]:
patients_copy.head()

patient_id assigned_sex  ...       phone_number              Email_address
0           1       female  ...       951-719-9170   ZoeWellish@superrito.com
1           2       female  ...  +1 (217) 569-3204       PamelaSHill@cuvox.de
2           3         male  ...       402-363-6804       JaeMDebord@gustr.com
3           4         male  ...  +1 (732) 636-8246  PhanBaLiem@jourrapide.com
4           5         male  ...       334-515-7487        TimNeudorf@cuvox.de

[5 rows x 16 columns]

In [152]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 486 entries, 0 to 502
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype   
---  ------         --------------  -----   
 0   patient_id     486 non-null    int64   
 1   assigned_sex   486 non-null    category
 2   given_name     486 non-null    object  
 3   surname        486 non-null    object  
 4   address        486 non-null    object  
 5   city           486 non-null    object  
 6   state          486 non-null    object  
 7   zip_code       486 non-null    object  
 8   country        486 non-null    object  
 9   contact        486 non-null    object  
 10  birthdate      486 non-null    object  
 11  weight         486 non-null    float64 
 12  height         486 non-null    int64   
 13  bmi            486 non-null    float64 
 14  phone_number   486 non-null    object  
 15  Email_address  486 non-null    object  
dtypes: category(1), float64(2), int64(2), object(11)
memory usage: 61.3+ KB


In [153]:
patients_copy.drop('contact',axis=1,inplace=True)

In [154]:
patients_copy.sample(5)

patient_id assigned_sex  ...  phone_number               Email_address
140         141         male  ...  217-421-0935  JohnTeichelmann@dayrep.com
436         437         male  ...  703-547-0551            SunKo@einrot.com
445         446         male  ...  408-792-9489    MaximusHenzen@einrot.com
476         477       female  ...  267-972-3749   BertaNapolitani@rhyta.com
453         454         male  ...  281-556-4376            WuSung@gustr.com

[5 rows x 15 columns]

 FINAL DATASETS AFTER DATA WRANGLING
 

In [155]:
patients_copy.sample(5)

patient_id assigned_sex  ...  phone_number                 Email_address
337         338         male  ...  208-830-2415  LeonReynosoRendon@einrot.com
52           53       female  ...  515-727-9324    BriannaLansell@fleckens.hu
324         325         male  ...  202-238-2247           KongLei@fleckens.hu
122         123         male  ...  231-607-3625         TaHaBoi@superrito.com
246         247       female  ...  714-431-2746          MeyaSjoberg@cuvox.de

[5 rows x 15 columns]

In [158]:
treatment_copy.sample(5)

given_name    surname  hba1c_start  ...  start  end adverse_reaction
299      Alvin    Jackson         7.62  ...     38   43                N
148     Benoît     Bonami         9.82  ...     44   43                N
332      Sofia  Hermansen         8.90  ...     34   34                Y
287       Dani     Antoun         7.73  ...     36   44                N
313      Žarka        Rap         7.54  ...     35   48                N

[5 rows x 9 columns]